### PR TITLE
BREAKING CHANGE: Use ESM, change scraper API, require DATA_DIR in env

### PR DESCRIPTION
**Pull request by sidvishnoi**
- Started using TypeScript for development ✨ .
- Migrate from commonjs to ESM ❇️ .
- Scraper is no longer a process. It maybe run in a Worker thread.
  ``` js
  import { main as scrapeCaniuse } from './scraper';
  const isDataUpdated = await scrapeCaniuse(); // Can avoid a cache purge
  ```
- enhancement(scraper): don't parse if data hasn't updated. 
- Environment variable `DATA_DIR`, whose value is the directory where
  data is saved, is required. This is done for consistency. This also
  avoids hard coding a directory in code.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import path from 'path';
```
Comment (sidvishnoi) [inline index.ts]: this file is same as before, except with added type info and use of `DATA_DIR`

File: scraper.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  const fileNames = await readdir(INPUT_DIR);
+  console.log(`Processing ${fileNames.length} files...`);
+  const promisesToProcess = fileNames.map(processFile);
+  await Promise.all(promisesToProcess);
+  console.log(`Processed ${fileNames.length} files.`);
+  return true;
+}
+
+function updateInputSource() {
```
Comment (sidvishnoi) [inline scraper.ts:50]: this is new. removes the shell script and runs git here.

File: scraper.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+interface Output {
+  [browserName: string]: [string, ReturnType<typeof formatStatus>][];
+}
+
+const INPUT_DIR = path.join(DATA_DIR, './caniuse-raw/features-json/');
+const OUTPUT_DIR = path.join(DATA_DIR, './caniuse/');
+
+export async function main() {
```
Comment (sidvishnoi) [inline scraper.ts:29]: this is updated as "no longer a process". This can be imported in respec.org (maybe in a Worker)

File: constants.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { resolve } from 'path';
+
+if (!process.env.DATA_DIR) {
+  throw new Error('DATA_DIR env variable must be set');
+}
+export const DATA_DIR = resolve(process.env.DATA_DIR);
+
+export const BROWSERS = new Set([
```
Comment (sidvishnoi) [inline constants.ts:8]: moved from `index.js`

File: package.json [hunk 5]
```diff
@@ -1,21 +1,24 @@
...
-    "fetch-caniuse-data": "./scraper/get-data.sh",
-    "convert-caniuse-data": "./scraper/process-data.js"
-  },
   "scripts": {
-    "fetch": "./scraper/get-data.sh",
-    "convert": "./scraper/process-data.js"
+    "build": "tsc -d",
+    "clean": "rm -rf *.js *.d.ts",
+    "dev": "tsc -w",
+    "prepare": "npm run clean && npm run build"
```
Comment (sidvishnoi) [inline package.json:20]: npm will still have only the JS version (and `.d.ts` files). We don't need to migrate respec.org to TypeScript

File: scraper.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    let hasUpdated = true;
+    git.stdout.on('data', (data: ArrayBuffer) => {
+      hasUpdated = !data.toString().includes('Already up to date');
+    });
+    git.on('error', (error: Error) => {
+      reject(error);
+    });
+    git.on('exit', (code: number) => {
+      if (code !== 0) {
+        reject(new Error(`The process existed with code ${code}`));
```
Comment (sidvishnoi) [inline scraper.ts]:
```suggestion
        reject(new Error(`The process exited with code ${code}`));
```

File: scraper.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    : ['pull', 'origin', 'master'];
+  const cwd = shouldClone ? path.resolve(DATA_DIR) : dataDir;
+
+  return new Promise<boolean>((resolve, reject) => {
+    const git = spawn('git', args, { cwd });
+    let hasUpdated = true;
+    git.stdout.on('data', (data: ArrayBuffer) => {
+      hasUpdated = !data.toString().includes('Already up to date');
+    });
+    git.on('error', (error: Error) => {
```
Comment (sidvishnoi) [inline scraper.ts]:
```suggestion
    git.on('error', reject);
```

File: scraper.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      reject(error);
```
Comment (sidvishnoi) [inline scraper.ts]:
```suggestion
```

File: scraper.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    });
```
Comment (sidvishnoi) [inline scraper.ts]:
```suggestion
```

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "target": "esnext",
+    "module": "esnext",
+    "esModuleInterop": true,
```
Comment (sidvishnoi) [inline tsconfig.json]:
```suggestion
```

File: tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,10 @@
+{
+  "compilerOptions": {
+    "target": "esnext",
+    "module": "esnext",
+    "removeComments": false,
+    "strict": true,
+    "newLine": "LF",
+    "allowSyntheticDefaultImports": true
```
Comment (sidvishnoi) [inline tsconfig.json]:
```suggestion
```

File: tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,10 @@
+{
+  "compilerOptions": {
+    "target": "esnext",
+    "module": "esnext",
+    "removeComments": false,
+    "strict": true,
+    "newLine": "LF",
```
Comment (sidvishnoi) [inline tsconfig.json]:
```suggestion
    "newLine": "LF"
```

**Review by marcoscaceres — APPROVED** [review]
Looks good.

**Pull request merged**